vllm/core/evictor.py

```diff
@@ -85,34 +89,46 @@ def evict(self) -> Tuple[int, int]:
         if len(self.free_table) == 0:
             raise ValueError("No usable cache memory left")
 
-        evicted_block, evicted_block_id = None, None
-        # The blocks with the lowest timestamps should be placed consecutively
-        # at the start of OrderedDict. Loop through all these blocks to
-        # find the one with maximum number of hashed tokens.
-        for _id, block in self.free_table.items():
-            if evicted_block is None:
-                evicted_block, evicted_block_id = block, _id
-                continue
-            if evicted_block.last_accessed < block.last_accessed:
-                break
-            if evicted_block.num_hashed_tokens < block.num_hashed_tokens:
-                evicted_block, evicted_block_id = block, _id
-
-        assert evicted_block is not None
-        assert evicted_block_id is not None
-        self.free_table.pop(evicted_block_id)
-
-        return evicted_block_id, evicted_block.content_hash
+        while self.priority_queue:
+            # Lazy deletion algorithm is applied.
+            last_accessed, _, block_id, content_hash = heapq.heappop(
+                self.priority_queue)
+            if (block_id in self.free_table and
+                    self.free_table[block_id].last_accessed == last_accessed):
+                self.free_table.pop(block_id)
+                return block_id, content_hash
```

Comment:
I'm a bit worry about this lazy deletion algorithm as it is pretty hard to understand for others and easy to introduce bugs in corner cases. Here are some possible questions people may ask by reading this code:
1. How a block in the heap not in the free table? A related question is why we need to cleanup the heap.
2. How a block in the heap and the free table could have different last access time?

Reply:
@comaniac Thank you for the valuable feedback.
I've added comments regarding the lazy deletion process.

I understand your concerns about the lazy deletion algorithm, as it shows O(n log n) time complexity when triggered. However, since outdated entries are also removed through heap pops, I believe cleanup is not an operation that happens frequently.

In fact, I also considered using doubly linked lists and dictionaries for this optimization. While these structures are generally O(1), I think that if the key value changes from being solely based on the last accessed time (which always increases), adding entries could then take O(n) time. That’s why I opted for a priority queue. Nevertheless, I acknowledge the concerns about lazy deletion holding outdated entries.

Reply:
Yes I used doubly linked list in v1 prefix caching and it works well, but it would be tedious for v0.

Reply:
Oh I see. I'll check the v1 implementation later as well.